apps/prairielearn/src/lib/server-jobs.types.ts

```diff
@@ -0,0 +1,35 @@
+import { z } from 'zod';
+
+import { type Job, JobSchema, JobSequenceSchema, UserSchema } from './db-types.js';
+
+const JobRowSchema = JobSchema.extend({
+  start_date_formatted: z.string().nullable(),
+  finish_date_formatted: z.string().nullable(),
+  user_uid: UserSchema.shape.uid.nullable(),
+  authn_user_uid: UserSchema.shape.uid.nullable(),
+});
+type JobRow = z.infer<typeof JobRowSchema>;
+
+export const JobSequenceWithJobsSchema = JobSequenceSchema.extend({
+  start_date_formatted: z.string().nullable(),
+  finish_date_formatted: z.string().nullable(),
+  user_uid: UserSchema.shape.uid.nullable(),
+  authn_user_uid: UserSchema.shape.uid.nullable(),
+  job_count: z.coerce.number(),
+  jobs: JobRowSchema.array(),
+});
+export type JobSequenceWithJobs = z.infer<typeof JobSequenceWithJobsSchema>;
+
+type JobWithToken = JobRow & { token: string };
+export type JobSequenceWithTokens = Omit<JobSequenceWithJobs, 'jobs'> & {
+  token: string;
+  jobs: JobWithToken[];
+};
+
+export type JobWithFormattedOutput = Omit<JobWithToken, 'output'> & {
+  output: string;
+  output_raw: Job['output'];
+};
+export type JobSequenceWithFormattedOutput = Omit<JobSequenceWithTokens, 'jobs'> & {
+  jobs: JobWithFormattedOutput[];
+};
```

Comment:
In theory, the "with formatted output" could be removed in place of formatting directly in the template, right? Same with token generation?

Reply:
Indeed, done in 65af26c. For tokens I'd prefer not to move them as that might create some dependency nightmares between assets and server-side code.

I also converted the formatted date stuff in 890874f. Do you want to review those changes?

Reply:
> For tokens I'd prefer not to move them as that might create some dependency nightmares between assets and server-side code.

Is your concern that we'd eventually want to use the job sequence templates directly on the client?

I'll take a look at your most recent changes now.

Reply:
> Is your concern that we'd eventually want to use the job sequence templates directly on the client?

Reviewing what I have in mind, it's not actually that bad of an issue. Currently the jobSequenceResults.ts asset imports a type from the JobSequenceResults.html.ts component. If tokens were generated in the component, this type would need to go to a different types.ts file.

That said, I still have a bit of reservations about where the line lies between computations allowed or not done in an html.ts component/page. Anything related to formatting is fair game, but once we start adding calculation of tokens this starts to move the line a bit. Not a big deal, but I assume we'd want to keep some of this formatting content clean of this kind of calculation.